### DISCUSSION TITLE
Clarification on how task hubs work

### DISCUSSION THREAD

**Discussion by SYOM29** [Azure/azure-functions-durable-extension#1946] [Q&A] [answered]
Hello,

We have 2 separate durable functions to manage 2 different flows in a single functions app (single functions app instance when deployed). From what I have found in docs, every durable function should have a separate task hub assigned to it if all of them use a single storage account. However, I haven't found any specifics about how task hubs will behave in such scenarios.

Questions:
1. I haven't found a way to specify a separate config for each task hub in host.json. That said using normal regular expression in host.json (like "%MyHub%") seems to apply that configuration to each task hub that matches that naming (I have tested this only in the local environment), however, this means that for single functions application each task hub will have the same host.json configuration at all times (or default configuration if the name does not match I guess). Is that done by design or maybe there is another way to solve this?

2. If there are multiple durable functions, what is a recommended way of managing their task hubs? To be honest, the motivation behind this question is the limitation described in question 1. If we can apply the single configuration to every durable function does that mean that we should try to have a separate instance for each durable function?

3. If we have multiple durable functions deployed to a single instance, each with its own task hub, how will this affect parameters configured in host.json(like maxConcurrentOrchestratorFunctions)? I mean, for example, if we have 2 durable functions with 2 different task hubs and both having maxConcurrentOrchestratorFunctions: 10 and they both are deployed to a single instance, does that mean they will share this config (if one has already started 7 orchestrators the other will just be able to start 3)? Or will each durable function be able to run up to 10 concurrent orchestrators (up to the 10x number of CPUs on the machine)?

4. Just out of curiosity, what will happen if multiple durable functions will use the same hub in the same storage while being in the same application instance?

In light of the ambiguities described above, maybe it would be good to add some further clarification on how task hubs work on https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-task-hubs?tabs=csharp page or some other page.

Thank you in advance

**Comment by cgillum** [comment, marked as answer]
The docs suggest having each function *app* be configured with its own task hub, not each *function*. Durable Functions only works correctly when multiple functions share the same task hub. This is mentioned in the task hub docs [here](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-task-hubs).

> Orchestrator, activity, and entity functions can only directly interact with each other when they belong to the same task hub.

It's _not_ recommended (or possible?) to have a single function app interacting with multiple task hubs except when binding a durable client to another task hub for cross-app interactions. The way to enable this isn't host.json, but rather specifying the name of the task hub in the [durable client binding](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-bindings?tabs=csharp#orchestration-client). I don't think we allow orchestrator and activity functions to be configured in separate task hubs - these trigger bindings don't expose any task hub configuration.

The configuration in host.json only applies to the single task hub used by orchestrator and activity functions across the app.

I hope this helps clarify things.

**Comment by SYOM29** [comment]
Thank you very much for your answer @cgillum. It answered all my questions
